cmd/kops/toolbox_instance_selector.go

```diff
@@ -0,0 +1,487 @@
+		MemoryRange:            commandline.IntRangeMe(flags[memory]),
+		VCpusToMemoryRatio:     commandline.Float64Me(flags[vcpusToMemoryRatio]),
+		CPUArchitecture:        commandline.StringMe(flags[cpuArchitecture]),
+		GpusRange:              commandline.IntRangeMe(flags[gpus]),
+		GpuMemoryRange:         commandline.IntRangeMe(flags[gpuMemoryTotal]),
+		PlacementGroupStrategy: commandline.StringMe(flags[placementGroupStrategy]),
+		UsageClass:             commandline.StringMe(flags[usageClass]),
+		EnaSupport:             commandline.BoolMe(flags[enaSupport]),
+		Burstable:              commandline.BoolMe(flags[burstSupport]),
+		Region:                 commandline.StringMe(region),
+		AvailabilityZones:      commandline.StringSliceMe(flags[availabilityZones]),
+		MaxResults:             commandline.IntMe(flags[maxResults]),
+		NetworkInterfaces:      commandline.IntRangeMe(flags[networkInterfaces]),
+		NetworkPerformance:     commandline.IntRangeMe(flags[networkPerformance]),
+		AllowList:              commandline.RegexMe(flags[allowList]),
+		DenyList:               commandline.RegexMe(flags[denyList]),
+		InstanceTypeBase:       commandline.StringMe(flags[instanceTypeBase]),
+		Flexible:               commandline.BoolMe(flags[flexible]),
+	}
+}
+
+func getInstanceSelectorOpts(commandline *cli.CommandLineInterface) InstanceSelectorOptions {
+	opts := InstanceSelectorOptions{}
+	flags := commandline.Flags
+	opts.NodeCountMin = int32(*commandline.IntMe(flags[nodeCountMin]))
+	opts.NodeCountMax = int32(*commandline.IntMe(flags[nodeCountMax]))
+	opts.InstanceGroupName = *commandline.StringMe(flags[igName])
+	opts.Output = *commandline.StringMe(flags[output])
+	opts.DryRun = *commandline.BoolMe(flags[dryRun])
+	opts.ClusterAutoscaler = *commandline.BoolMe(flags[clusterAutoscaler])
+	if flags[nodeVolumeSize] != nil {
+		volumeSize := int32(*commandline.IntMe(flags[nodeVolumeSize]))
+		opts.NodeVolumeSize = &volumeSize
+	}
+	if flags[nodeSecurityGroups] != nil {
+		opts.NodeSecurityGroups = *commandline.StringSliceMe(flags[nodeSecurityGroups])
+	}
+	if flags[instanceGroupCount] != nil {
+		opts.InstanceGroupCount = *commandline.IntMe(flags[instanceGroupCount])
+	}
+	return opts
+}
+
+func getClusterZones(subnets []kops.ClusterSubnetSpec) ([]string, error) {
+	region := ""
+	zones := []string{}
+	for _, subnet := range subnets {
+		zoneRegion := subnet.Zone[:len(subnet.Zone)-1]
+		zones = append(zones, subnet.Zone)
+		if region != "" && zoneRegion != region {
+			return nil, fmt.Errorf("clusters cannot span multiple regions")
+		}
+		region = zoneRegion
+	}
+	if len(zones) == 0 {
+		return nil, fmt.Errorf("the cluster must include at least 1 subnet")
+	}
+	return zones, nil
+}
+
+func createInstanceGroup(groupName, clusterName string, zones []string) *kops.InstanceGroup {
+	ig := &kops.InstanceGroup{}
+	ig.ObjectMeta.Name = groupName
+	ig.Spec.Role = kops.InstanceGroupRoleNode
+	ig.Spec.Subnets = zones
```

Comment:
This has to be a list of subnets, not AZs. I think it may also be a case of public or private, would not be able to use both at same time.

Reply:
Looks like the current create_ig implementation is wrong then: https://github.com/kubernetes/kops/blob/bc074e857cef9facc96d0ee5dc25066d1ef3bdc1/cmd/kops/create_ig.go#L106-L107
https://github.com/kubernetes/kops/blob/bc074e857cef9facc96d0ee5dc25066d1ef3bdc1/cmd/kops/create_ig.go#L161-L162

Reply:
Actually, after I changed the implementation to look at the name of the subnet from the cluster I found that subnet.Name in ClusterSubnetSpec is the zone name (us-east-1a). I guess it's still better adaptable to changes if it's using the subnet name rather than relying on the zone name matching.

Reply:
Subnets can also have the "utility" prefix. public ones are named like "utility-<AZ>", the private ones are named as the AZ. At least these are the defaults.
https://github.com/kubernetes/kops/blob/master/docs/single-to-multi-master.md#create-new-subnets

Reply:
would you recommend changing `--zones` to `--subnets` to match `create ig` and allow users to pass in things like "utility-us-east-1a"  or "us-east-1a" and then add validation in instance-selector to make sure the subnets passed are all public or all private?

Reply:
I think it would probably be best that way.

Reply:
cool, I added the validation and updated to use subnets rather than zones 👍